genienlp/paraphrase/transformers_utils.py

```diff
@@ -63,1096 +84,257 @@
+    '''
+    vocab_files_names = {"vocab_file": "sentencepiece.bpe.model"}
+    max_model_input_sizes = {m: 1024 for m in _all_mbart_models}
+    pretrained_vocab_files_map = {"vocab_file": {m: SPM_URL for m in _all_mbart_models}}
 
-        self.init_weights()
+    prefix_tokens: List[int] = []
+    suffix_tokens: List[int] = []
 
+    def __init__(self, *args, tokenizer_file=None, **kwargs):
+        super().__init__(*args, tokenizer_file=tokenizer_file, **kwargs)
 
-class BartForConditionalGeneration(BartForConditionalGeneration):
-    base_model_prefix = "model"
+    def set_src_lang_special_tokens(self, src_lang) -> None:
+        """Reset the special tokens to the source lang setting. Prefix [bos_token_id], suffix =[eos_token_id]."""
+        self.cur_lang_code = self.lang_code_to_id[src_lang]
+        self.prefix_tokens = [self.bos_token_id]
+        self.suffix_tokens = [self.eos_token_id]
 
-    def __init__(self, config: BartConfig):
-        super().__init__(config)
-        base_model = BartModel(config)
-        self.model = base_model
-        self.register_buffer("final_logits_bias", torch.zeros((1, self.model.shared.num_embeddings)))
+    def set_tgt_lang_special_tokens(self, lang: str) -> None:
+        """Reset the special tokens to the target language setting. Prefix [tgt_lang_code], suffix =[eos_token_id]."""
+        self.cur_lang_code = self.lang_code_to_id[lang]
+        self.prefix_tokens = [self.cur_lang_code]
+        self.suffix_tokens = [self.eos_token_id]
         
-    def prepare_inputs_for_generation(self, decoder_input_ids, past, attention_mask, use_cache, **kwargs):
-        assert past is not None, "past has to be defined for encoder_outputs"
-
-        # first step, decoder_cached_states are empty
-        # first step
-        if kwargs['cur_len'] == 1:
-            encoder_outputs, decoder_cached_states = past[0], None
-        else:
-            if use_cache:
-                if len(past) < 2:
-                    encoder_outputs, decoder_cached_states = past[0], None
-                else:
-                    encoder_outputs, decoder_cached_states = past[0], past[1]
-            else:
-                encoder_outputs, decoder_cached_states = past[0], None
-                
-        if not isinstance(encoder_outputs, tuple):
-            encoder_outputs = (encoder_outputs, )
+###############
 
-        return {
-            "input_ids": None,  # encoder_outputs is defined. input_ids not needed
-            "encoder_outputs": encoder_outputs,
-            "decoder_cached_states": decoder_cached_states,
-            "decoder_input_ids": decoder_input_ids,
-            "attention_mask": attention_mask,
-            "use_cache": use_cache,  # change this to avoid caching (presumably for debugging)
-        }
 
-    def _generate_no_beam_search(
+class GeniePreTrainedModel(PreTrainedModel):
+    '''
+    General class for PreTrainedModel which can output cross-attention weights during generation
+    '''
```

Comment:
I thought cross-attention weights were natively supported now?

Reply:
The models output cross-attention in their `forward()` method, but running `generation()` still doesn't return them. Maybe we should make a PR for this.

Reply:
Please do! Let's get rid of this code...

Reply:
One thing though is that we need a subclass like this for returning other useful stuff such as confidence scores that `generation()` doesn't return. @s-jse may want to speak of his experience so far :)

Reply:
Yeah, we need a `GeniePreTrainedModel` class, we could get rid of the rest I think (unless a reason for model-specific behavior comes up in the future). Specifically, I need the generation code to output logits, and potentially other useful things that help us determine the model confidence.